src/compiler/perfLogger.ts

```diff
@@ -0,0 +1,81 @@
+/* @internal */
+namespace ts {
+    export type PerfLogger = typeof import("@microsoft/typescript-etw"); // tslint:disable-line:no-implicit-dependencies
+
+    export class NullLogger implements PerfLogger {
+        logEvent(_msg: string): void {
+            return;
+        }
+        logErrEvent(_msg: string): void {
+            return;
+        }
+        logPerfEvent(_msg: string): void {
+            return;
+        }
+        logInfoEvent(_msg: string): void {
+            return;
+        }
+        logStartCommand(_command: string, _msg: string): void {
+            return;
+        }
+        logStopCommand(_command: string, _msg: string): void {
+            return;
+        }
+        logStartUpdateProgram(_msg: string): void {
+            return;
+        }
+        logStopUpdateProgram(_msg: string): void {
+            return;
+        }
+        logStartUpdateGraph(): void {
+            return;
+        }
+        logStopUpdateGraph(): void {
+            return;
+        }
+        logStartResolveModule(_name: string): void {
+            return;
+        }
+        logStopResolveModule(_success: string): void {
+            return;
+        }
+        logStartParseSourceFile(_filename: string): void {
+            return;
+        }
+        logStopParseSourceFile(): void {
+            return;
+        }
+        logStartReadFile(_filename: string): void {
+            return;
+        }
+        logStopReadFile(): void {
+            return;
+        }
+        logStartBindFile(_filename: string): void {
+            return;
+        }
+        logStopBindFile(): void {
+            return;
+        }
+        logStartScheduledOperation(_operationId: string): void {
+            return;
+        }
+        logStopScheduledOperation(): void {
+            return;
+        }
+    }
+
+    // Load optional module to enable Event Tracing for Windows
+    // See https://github.com/microsoft/typescript-etw for more information
+    let etwModule;
+    try {
+        // require() will throw an exception if the module is not installed
+        // It may also return undefined if not installed properly
+        etwModule = require("@microsoft/typescript-etw"); // tslint:disable-line:no-implicit-dependencies
+    }
+    catch (e) {
+        etwModule = undefined;
+    }
+
+    export const perfLogger: PerfLogger = etwModule ? etwModule : new NullLogger();

```

Comment:
> I considered the second suggestion (moving conditional logic inside the logger class), but that is very unwieldy given the strong typing that is currently in place. We would need logInfoEventIf(), logStartCommandIf(), logStopCommandIf(), etc... It seems like overkill for this problem.

Here's a version using a fluent API to avoid remaking conditional methods with similar definitions:

```suggestion
    const nullLogger = new NullLogger();
    class PassThruLogger implements PerfLogger {
        constructor(protected base: PerfLogger) {}
        logEvent(msg: string): void {
            return this.base.logEvent(msg);
        }
        logErrEvent(msg: string): void {
            return this.base.logErrEvent(msg);
        }
        logPerfEvent(msg: string): void {
            return this.base.logPerfEvent(msg);
        }
        logInfoEvent(msg: string): void {
            return this.base.logInfoEvent(msg);
        }
        logStartCommand(command: string, msg: string): void {
            return this.base.logStartCommand(command, msg);
        }
        logStopCommand(command: string, msg: string): void {
            return this.base.logStopCommand(command, msg);
        }
        logStartUpdateProgram(msg: string): void {
            return this.base.startUpdateProgram(msg);
        }
        logStopUpdateProgram(msg: string): void {
            return this.base.logStopUpdateProgram(msg);
        }
        logStartUpdateGraph(): void {
            return this.base.logStartUpdateGraph();
        }
        logStopUpdateGraph(): void {
            return this.base.logStopUpdateGraph();
        }
        logStartResolveModule(name: string): void {
            return this.base.logStartResolveModule(name);
        }
        logStopResolveModule(_success: string): void {
            return this.base.logStopResolveModule(success);
        }
        logStartParseSourceFile(filename: string): void {
            return this.base.logStartParseSourceFile(filename);
        }
        logStopParseSourceFile(): void {
            return this.base.logStopParseSourceFile();
        }
        logStartReadFile(filename: string): void {
            return this.base.logStartReadFile(filename);
        }
        logStopReadFile(): void {
            return this.base.logStopReadFile();
        }
        logStartBindFile(filename: string): void {
            return this.base.logStartBindFile(filename);
        }
        logStopBindFile(): void {
            return this.base.logStopBindFile();
        }
        logStartScheduledOperation(operationId: string): void {
            return this.base.logStartScheduledOperation(operationId);
        }
        logStopScheduledOperation(): void {
            return this.base.logStopScheduledOperation();
        }
    }
    class ConditionalPerfLogger extends PassThruLogger {
        constructor(base: PerfLogger) {
            super(base);
        }
        if(condition: boolean): PerfLogger {
          if (condition) {
            return this.base;
          }
          return nullLogger;
        }
    }
    export const perfLogger: ConditionalPerfLogger = new ConditionalPerfLogger(etwModule ? etwModule : nullLogger);
```
usage:
```
perfLogger.if(moduleResolution).logEvent("message");
```

Reply:
I tried this out but it's still missing the piece where the message is a function instead of a string. Without that the string interpolation could fail with a null reference.

At this point I'd like to move forward with the code that I have which is pretty straightforward and has addressed all of the other comments. What do you think @sheetalkamat ?

Reply:
You can redefine `if` to be
```ts
        if(condition: boolean, cb: (logger: PerfLogger) => void): void {
          if (condition) {
            return cb(this.base);
          }
        }
```
with usage
```
perfLogger.if(moduleResolution, log => log.logEvent("message"));
```
if that's the concern.